Add hex input

kfox opened the pull request:
@ustudio/dev

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ustudio/paper-color-picker/1)
<!-- Reviewable:end -->

Commit:
work in progress

Commit:
continued refactoring to reduce redundancy

Commit:
centralized common color behaviors

Commit:
mostly working

Commit:
more tests

Commit:
added more tests

Commit:
change name of behavior

Commit:
update bower.json

Commit:
accept changes via enter key, allow configurable button text

Commit:
add test for enter key

Commit:
test cleanup

Commit:
fixed most issues, still need to fix a couple minor ones and flesh out documentation

Commit:
more cleanup

Commit:
improve documentation

Commit:
code cleanup

Commit:
more fixes

Commit:
still need to fix the callback.call error

Commit:
use debounce correctly

Commit:
use tinycolor instead of built-in conversion functions

Commit:
minor fixes

Commit:
almost there

Commit:
one bug left

Commit:
fixed but tests still failing

Commit:
fixed tests

fearphage commented:
Several small things and a few questions and clarifications. 

A lot of the basic formatting stuff would have been caught (and fixed) by a liberal use of `ESLint` (https://github.com/ustudio/eslint-config-ustudio). @kfox 

---

Reviewed 19 of 27 files at r1.
Review status: 19 of 27 files reviewed at latest revision, 23 unresolved discussions.

---

*[bower.json, line 3 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7V6q99asgYf-G9V-S:-KZ7V6q99asgYf-G9V-T:b-chhvwi) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/bower.json#L3)):*
> ```JSON
> {
>   "name": "paper-color-picker",
>   "homepage": "https://github.com/ustudio/paper-color-picker",
> ```

Do we want to take over this project?

---

*[bower.json, line 7 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7VBLdSN9NecRQqJu-:-KZ7VBLdSN9NecRQqJu0:bjcvigi) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/bower.json#L7)):*
> ```JSON
>     "David Mulder",
>     "Dennis Kugelmann",
>     "Kelly Fox"
> ```

No individual names.

One band. One sound. -- https://youtu.be/ZnAnGsrv4dU

---

*[hex-input.html, line 97 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7YxF0oVZcclayJJmM:-KZ7YxF0oVZcclayJJmN:b-ukgspn) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/hex-input.html#L97)):*
> ```HTML
>       if (hex[0] !== '#' || hex.indexOf('#', 1) > -1) {
>         hex = hex.replace(/#/g, '');
>         this.set('value', '#' + hex);
> ```

The output passed to `this.set` is inconsistent. Here it could be lowercased hex. A few lines down you convert it to uppercase. A few lines after that you don't do anything with the case again. Is the case of the string important?

It seems like you want several if statements and only one call to `this.set` at the end.

---

*[hex-input.html, line 107 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7YKaPSlOLzitXj7iz:-KZ7YKaPSlOLzitXj7j-:bqzyge2) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/hex-input.html#L107)):*
> ```HTML
> 
>       if (hex.length > 7) {
>         this.set('value', '#' + hex.sub(1, 7));
> ```

Is `hex` just a string? `String.prototype.sub` is not a thing. `substr` and `substring` exist. 

---

*[hex-input.html, line 110 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZAPrCgvi3jzw0wgC:-KZ7ZAPrCgvi3jzw0wgD:bv9mhu5) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/hex-input.html#L110)):*
> ```HTML
>       };
> 
>       return;
> ```

Remove. This isn't doing anything.

---

*[paper-color-circle.html, line 100 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7Pkzx1pfcDbsK1AoG:-KZ7Pkzx1pfcDbsK1AoH:b-lgxmg5) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L100)):*
> ```HTML
>         },
> 
>         _size: {
> ```

MINOR: `_size: Number`

---

*[paper-color-circle.html, line 117 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7PuM8eHMJWtGyQD6b:-KZ7PuM8eHMJWtGyQD6c:b-aqh4vt) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L117)):*
> ```HTML
> 
>       listeners: {
>         'iron-resize': '_onResize'
> ```

Did this change get us anything? It looks like it's only used here and all it does is call redraw like the old code. Revert?

---

*[paper-color-circle.html, line 138 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7QLikwm_l2T1NZgsh:-KZ7QLikwm_l2T1NZgsi:bxznckq) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L138)):*
> ```HTML
>           );
> 
>           if (distance < this._size / 100) return;
> ```

MINOR: Bring back the braces.

---

*[paper-color-circle.html, line 178 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7QXmsKTzmGwIhWt4b:-KZ7QXmsKTzmGwIhWt4c:b-iimpyo) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L178)):*
> ```HTML
>        * Force the circle to be redrawn.
>        */
>       redraw: function (e, detail) {
> ```

Are you using either of these variables? I don't see it.

---

*[paper-color-circle.html, line 284 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RDsCItKdSBSC_V4O:-KZ7RDsCItKdSBSC_V4P:bc7iw9k) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L284)):*
> ```HTML
>       },
> 
>       _brightnessChanged: function (brightness) {
> ```

Why not just call redraw directly?

---

*[paper-color-circle.html, line 288 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RD5rHpVmdWgijqwC:-KZ7RD5rHpVmdWgijqwD:bc7iw9k) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L288)):*
> ```HTML
>       },
> 
>       _luminosityChanged: function (luminosity) {
> ```

Why not just call redraw directly?

---

*[paper-color-input.html, line 95 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RxX1uZFqCLOhIUVx:-KZ7RxX1uZFqCLOhIUVy:b-yqisl4) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/paper-color-input.html#L95)):*
> ```HTML
> 				 *
> 				 * @attribute label
> 				 * @type String
> ```

Why'd you dump the JSDoc strings?

---

*[paper-color-input.html, line 184 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7S1jRxn28Sqe8pY4z:-KZ7S1jRxn28Sqe8pY5-:b-k343jj) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-input.html#L184)):*
> ```HTML
>         if (this.hex !== undefined) {
>           this._onHexChanged(this.hex);
>         } else if (this.red !== undefined &&
> ```

cuddled else's

---

*[paper-color-input.html, line 214 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7SIGj4FFugfVPywQU:-KZ7SIGj4FFugfVPywQV:b-3xz7ru) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-input.html#L214)):*
> ```HTML
> 
>       _computeDescription: function (hex, name, useHexAsTitle) {
>         return useHexAsTitle ? name : hex;
> ```

`useHexAsTitle` doesn't make sense here. `useNameAsDescription`?

---

*[paper-color-picker-behavior.html, line 3402 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7T0yJbI28uQRzdpHv:-KZ7T0yKk1AH0bt4cmYS:bwq8oq7) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker-behavior.html#L3402)):*
> ```HTML
>         * The red component of the current RGB value.
>         */
>       red: {
> ```

MINOR: Simplify? `blah: Number` and `foo: String` throughout

---

*[paper-color-picker-behavior.html, line 3467 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7TASJ0oE0Oi6ds3Q0:-KZ7TASJ0oE0Oi6ds3Q1:b-d92pcd) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker-behavior.html#L3467)):*
> ```HTML
>         * code as the description.
>         */
>       useHexAsTitle: {
> ```

The property says `title` while the comment says `description`. Who is correct?

---

*[paper-color-picker.html, line 480 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7TylVKj4MgEjVECNN:-KZ7TylVKj4MgEjVECNO:b10hpl8) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker.html#L480)):*
> ```HTML
>         if (this.hex !== undefined) {
>           this._onHexChanged(this.hex);
>         } else if (this.red !== undefined &&
> ```

No cuddling.

---

*[paper-color-picker.html, line 553 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7U76BdIz1tv6p9zqN:-KZ7U76BdIz1tv6p9zqO:b5j0ip6) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker.html#L553)):*
> ```HTML
>         var rgb = { r: this.red, g: this.green, b: this.blue };
>         this.fire('color-picked', rgb);
>         this.fire('update-color', rgb);
> ```

Why are there 2 differently named events firing with the same data?

---

*[demo/favicon.ico, line 0 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7VYA8DR8OYgrThoLt:-KZ7VYA8DR8OYgrThoLu:bzgs64x) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/demo/favicon.ico#L0)):*
How did this empty file change?

Should it be deleted?

---

*[test/hex-validatable.html, line 5 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7WtJ2qqO0pw_tC9pZ:-KZ7WtJ3Z7BOhJPpLxnH:b-pcf4t1) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/hex-validatable.html#L5)):*
> ```HTML
> <link rel="import" href="../../iron-validatable-behavior/iron-validatable-behavior.html">
> 
> <dom-module id="hex-validatable">
> ```

MINOR: `validatable hex` sounds more natural/normal to say. Any reason you named it this way?

---

*[test/paper-color-circle.html, line 0 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZTm3ctY-D_BE-mxR:-KZ7ZTm4A_5f_hRliit7:b6src2q) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-circle.html#L0)):*
Missing tests

---

*[test/paper-color-input.html, line 0 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7Z_LSKAtRYWvLLRHx:-KZ7Z_LSKAtRYWvLLRHy:b6src2q) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-input.html#L0)):*
Missing tests

---

*[test/rgb-component-validatable.html, line 12 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZkuE_vGyF7y-_6o5:-KZ7ZkuE_vGyF7y-_6o6:b6src2q) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/rgb-component-validatable.html#L12)):*
> ```HTML
> 
> <script>
>   Polymer({
> ```

Missing tests

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-:-KZ7Y_-ZOulnkidgkIMo:bf53nd3)*
<!-- Sent from Reviewable.io -->

tjensen commented:
Reviewed 4 of 27 files at r1.
Review status: 21 of 27 files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

tjensen commented:
*[test/paper-color-picker-behavior.html, line 72 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7jLu3vZx38BtSn3wk:-KZ7jLu3vZx38BtSn3wl:b-iinwth) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L72)):*
> ```HTML
>           assert.equal(element.green, 83);
>           assert.equal(element.blue, 35);
>           assert.equal(element.hue, 14.117647058823529);
> ```

MINOR: Is it safe to do an exact equals comparison on a floating point value, like this? I'm used to having to do checks that a value is within some range of accuracy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

tjensen commented:
*[test/paper-color-picker.html, line 38 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7m5V152Z-BAzix_CN:-KZ7m5V2aio_bpNQSd9L:b-ki7c0e) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L38)):*
> ```HTML
>       this.timeout(2500);
> 
>       var element;
> ```

I've been informed that the convention we've been using is to make these things like, `this.element`, instead of defining them as `var`iables. There might be some small benefits to doing it that way?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

tjensen commented:
Reviewed 3 of 27 files at r1.
Review status: 24 of 27 files reviewed at latest revision, 25 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

tjensen commented:
Are `test/paper-color-circle.html` and `test/paper-color-input.html` intentionally left blank?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-:-KZ7nd_JrO9o3RjZ2r2F:bxmkqpk)*
<!-- Sent from Reviewable.io -->

kfox commented:
@fearphage This repo is public and that repo is private. :(
@tjensen They were placeholders and need to be fleshed out. The primary goal was to get the color picker working and test the critical parts. I think we'll need to add a few more tests, but likely that will be after deploying this initial rewrite.

---

Review status: 24 of 27 files reviewed at latest revision, 25 unresolved discussions.

---

*[bower.json, line 3 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7V6q99asgYf-G9V-S:-KZ7adO4Yoetis2CK6V8:b-hhusng) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/bower.json#L3)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Do we want to take over this project?
</blockquote></details>

The original author hasn't replied to a direct message from me and he hasn't touched the repo in about a year. I'm thinking this is abandoned. So yeah, I guess we will.

---

*[bower.json, line 7 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7VBLdSN9NecRQqJu-:-KZ7avtTMIZ8b_fNwRHS:b-vprbsi) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/bower.json#L7)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

No individual names.

One band. One sound. -- https://youtu.be/ZnAnGsrv4dU
</blockquote></details>

I'll just put `uStudio, Inc.` in there.

---

*[hex-input.html, line 97 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7YxF0oVZcclayJJmM:-KZ83YqJbcFgBYv8B9EI:b2ewn03) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/hex-input.html#L97)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

The output passed to `this.set` is inconsistent. Here it could be lowercased hex. A few lines down you convert it to uppercase. A few lines after that you don't do anything with the case again. Is the case of the string important?

It seems like you want several if statements and only one call to `this.set` at the end.
</blockquote></details>

Refactoring.

---

*[hex-input.html, line 107 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7YKaPSlOLzitXj7iz:-KZ83Q8wmOokHO4yKcjE:b3ta79y) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/hex-input.html#L107)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Is `hex` just a string? `String.prototype.sub` is not a thing. `substr` and `substring` exist. 
</blockquote></details>

Ah, that was a sloppy mistake. Thanks!

---

*[paper-color-circle.html, line 117 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7PuM8eHMJWtGyQD6b:-KZ7e_bcHtHS1_NlVAvP:b9cpab) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L117)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Did this change get us anything? It looks like it's only used here and all it does is call redraw like the old code. Revert?
</blockquote></details>

It was useful in debugging, but yeah it doesn't buy anything.

---

*[paper-color-circle.html, line 178 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7QXmsKTzmGwIhWt4b:-KZ7fVDUWf1c5XGLYsd0:b4nq7jv) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L178)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Are you using either of these variables? I don't see it.
</blockquote></details>

Good catch. I was using them during debugging, but they're not needed now.

---

*[paper-color-circle.html, line 284 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RDsCItKdSBSC_V4O:-KZ7fmyF3fXs1h5fzrdu:brya9o) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L284)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Why not just call redraw directly?
</blockquote></details>

I'm using an observer instead of a listener, and am not auto-notifying (to prevent multiple color updates between elements).

---

*[paper-color-circle.html, line 288 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RD5rHpVmdWgijqwC:-KZ7g3YVG_p6SclHgglt:b-5s0dll) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L288)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Why not just call redraw directly?
</blockquote></details>

I'm using an observer instead of a listener, and am not auto-notifying (to prevent multiple color updates bouncing between elements).

---

*[paper-color-input.html, line 95 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RxX1uZFqCLOhIUVx:-KZ7nYkAAHWumZ1c70a8:bim1kxq) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/paper-color-input.html#L95)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Why'd you dump the JSDoc strings?
</blockquote></details>

It was an overly-aggressive attempt to cleanup the code. I will add them back in.

---

*[paper-color-input.html, line 214 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7SIGj4FFugfVPywQU:-KZ7mWbxbKNyzJ8ybHym:b-mp18jp) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-input.html#L214)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

`useHexAsTitle` doesn't make sense here. `useNameAsDescription`?
</blockquote></details>

I think that would add unnecessary complexity to do so. I'd have to add a new property to support that, as well as a new function to compute the value of that property, whereas the `useHexAsTitle` property is already present as part of the imported color picker behavior.

I think you can look at it like this: if we're using the hex code as the title of the color input and color picker, then the computed value of the description is just the color name. Otherwise, the description is the hex code.

---

*[paper-color-picker-behavior.html, line 3467 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7TASJ0oE0Oi6ds3Q0:-KZ7hxotdQBUF9aGBJDg:b-z0z8ul) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker-behavior.html#L3467)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

The property says `title` while the comment says `description`. Who is correct?
</blockquote></details>

Both are correct, but maybe the comment could be clearer. Something like: `When this property is true, the title is the hex value and the color name becomes the description, and vice versa if the property is false.`

---

*[paper-color-picker.html, line 553 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7U76BdIz1tv6p9zqN:-KZ7oTOwSipX7R7-aIiG:b-uyzcb4) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker.html#L553)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Why are there 2 differently named events firing with the same data?
</blockquote></details>

Good catch. At one point, the values being sent were different.

---

*[demo/favicon.ico, line  at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7VYA8DR8OYgrThoLt:-KZ85alH68siWUFQ7l9U:bcpcow1) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/demo/favicon.ico#L)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

How did this empty file change?

Should it be deleted?
</blockquote></details>

Deleting.

---

*[test/hex-validatable.html, line 5 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7WtJ2qqO0pw_tC9pZ:-KZ7p-ooctmT7umWkiqA:bw4pr2n) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/hex-validatable.html#L5)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

MINOR: `validatable hex` sounds more natural/normal to say. Any reason you named it this way?
</blockquote></details>

To conform with the conventions I saw in other Polymer elements. It looked to me like other validators had `<type-of-thing>-validatable` elements as part of their test suites, too.

---

*[test/paper-color-circle.html, line  at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZTm3ctY-D_BE-mxR:-KZ85eE3fMGuzqUARlYm:b-jbrnnw) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-circle.html#L)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Missing tests
</blockquote></details>

Removing. Will need to add this later.

---

*[test/paper-color-input.html, line  at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7Z_LSKAtRYWvLLRHx:-KZ85hk3NpnB-ISI1kD4:b-jbrnnw) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-input.html#L)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Missing tests
</blockquote></details>

Removing. Will need to add this later.

---

*[test/paper-color-picker-behavior.html, line 72 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7jLu3vZx38BtSn3wk:-KZ80fhCkFumQ07SyYgt:bq6wjtj) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L72)):*
<details><summary><i>Previously, tjensen (Tim Jensen) wrote…</i></summary><blockquote>

MINOR: Is it safe to do an exact equals comparison on a floating point value, like this? I'm used to having to do checks that a value is within some range of accuracy.
</blockquote></details>

Excellent point. I'll use the `approximately` assertion instead of `equal`.

---

*[test/paper-color-picker.html, line 38 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7m5V152Z-BAzix_CN:-KZ82xSpvpmQp7wcfHYE:b-kbxi9o) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L38)):*
<details><summary><i>Previously, tjensen (Tim Jensen) wrote…</i></summary><blockquote>

I've been informed that the convention we've been using is to make these things like, `this.element`, instead of defining them as `var`iables. There might be some small benefits to doing it that way?
</blockquote></details>

This follows the Polymer conventions that I've seen, and I'm not sure we'd get any benefit in this case since we're not dealing with objects as much as function blocks. If there are other benefits, e.g. performance/memory, I'm not sure.

---

*[test/rgb-component-validatable.html, line 12 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZkuE_vGyF7y-_6o5:-KZ85pxQi-lXpgn5xYwN:bif3n0f) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/rgb-component-validatable.html#L12)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Missing tests
</blockquote></details>

No test required here. This is only an element used by the `rgb-component-validator` test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-:-KZ8434YH90DlggY_G49:b-gp0p5d)*
<!-- Sent from Reviewable.io -->

kfox commented:
Review status: 24 of 27 files reviewed at latest revision, 25 unresolved discussions.

---

*[hex-input.html, line 110 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZAPrCgvi3jzw0wgC:-KZ87vIkB5UnHLzCBn7c:b-896fix) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/hex-input.html#L110)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Remove. This isn't doing anything.
</blockquote></details>

Done.

---

*[paper-color-circle.html, line 100 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7Pkzx1pfcDbsK1AoG:-KZ87wqu71rIaO_2t7CG:b-896fix) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L100)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

MINOR: `_size: Number`
</blockquote></details>

Done.

---

*[paper-color-circle.html, line 138 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7QLikwm_l2T1NZgsh:-KZ87xTlDzlMCUmVsnBC:b-896fix) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-circle.html#L138)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

MINOR: Bring back the braces.
</blockquote></details>

Done.

---

*[paper-color-input.html, line 184 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7S1jRxn28Sqe8pY4z:-KZ87y0dmSJ3BJ08a2d9:bt2vr1b) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-input.html#L184)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

cuddled else's
</blockquote></details>

Fixed.

---

*[paper-color-picker-behavior.html, line 3402 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7T0yJbI28uQRzdpHv:-KZ88-Xrzs6IjbCD-U1E:b-896fix) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker-behavior.html#L3402)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

MINOR: Simplify? `blah: Number` and `foo: String` throughout
</blockquote></details>

Done.

---

*[paper-color-picker.html, line 480 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7TylVKj4MgEjVECNN:-KZ880AHcO2c5XEA8Xwv:bt2vr1b) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-picker.html#L480)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

No cuddling.
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

fearphage commented:
Reviewed 6 of 27 files at r1.
Review status: all files reviewed at latest revision, 35 unresolved discussions.

---

*[test/paper-color-circle.html, line 0 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZDXt7hSrahaXteLP07:-KZDXt7iwMK-aempRW8B:bqw8aai) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-circle.html#L0)):*
What's up with the empty files? Are tests missing?

---

*[test/paper-color-input.html, line 0 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZDXx8UEhx_hw2ViTbF:-KZDXx8VY_vL5mh_KEac:bqw8aai) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-input.html#L0)):*
What's up with the empty files? Are tests missing?

---

*[test/paper-color-picker-behavior.html, line 41 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9yf6Oy0UHP1Qe9zGi:-KZ9yf6Oy0UHP1Qe9zGj:bgfqojz) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L41)):*
> ```HTML
>         Polymer({
>           is: 'x-behavior-test',
>           behaviors: [ PaperColorPicker.ColorBehavior ]
> ```

This is the only instance that I've seen where this array is not spread across 3 lines.

---

*[test/paper-color-picker-behavior.html, line 64 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9z-DrRuShI6Ro3v2s:-KZ9z-DrRuShI6Ro3v2t:bad60e9) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L64)):*
> ```HTML
>         suiteSetup(function () {
>           element = fixture('rgb');
>           element._onRGBChanged(element.red, element.green, element.blue);
> ```

Is this necessary?

---

*[test/paper-color-picker-behavior.html, line 69 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ySi_X7Ozq_tvqF06:-KZ7ySi_X7Ozq_tvqF07:bytb26) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L69)):*
> ```HTML
>         test('values are initialized properly', function () {
>           assert.equal(element.hex, '#EF5323');
>           assert.equal(element.red, 239);
> ```

It would be more meaningful to me as the reader to know where this value came from (IE. do the calulation here). The same is true for all these magic values.

````
red === hex.redPart().convertedToDecimal() // pseduo code, but just to illustrate the point
```

---

*[test/paper-color-picker-behavior.html, line 73 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9xTZh8Lvxpqy90Zfq:-KZ9xTZiBGG-31jhhE9S:bpt9n5v) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L73)):*
> ```HTML
>           assert.equal(element.blue, 35);
>           assert.equal(element.hue, 14.117647058823529);
>           assert.equal(element.saturation, 0.864406779661017);
> ```

Where do these values come from?

---

*[test/paper-color-picker-behavior.html, line 99 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9xtjNdZrfvarjnmIy:-KZ9xtjNdZrfvarjnmIz:b-14y7pp) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L99)):*
> ```HTML
>           assert.equal(element.red, 0xEF);
>           assert.equal(element.blue, 0x23);
>           assert.equal(element.hue, 64.36363636363637);
> ```

Where's this come from? Why is it this value? This is just a random number to me. 

This assertion doesn't have any value. If I made some changes and `element.hue` was equal to `68` instead, I would have no clue why it was broken. Make these values mean something.

---

*[test/paper-color-picker.html, line 30 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9sXYJcyDkQA3DAwJP:-KZ9sXYJcyDkQA3DAwJQ:bj4454) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L30)):*
> ```HTML
>   <script>
> 
>     function runAfterOpen (dialog, callback) {
> ```

Add this to the `beforeEach`.

---

*[test/paper-color-picker.html, line 36 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9sRma26lrd52zMsct:-KZ9sRma26lrd52zMscu:blt77l5) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L36)):*
> ```HTML
> 
>     suite('state management', function () {
>       this.timeout(2500);
> ```

What's so slow here?

---

*[test/paper-color-picker.html, line 43 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ8hy4_7QUNk6GWGW0Z:-KZ8hy4ayLTt5KVsgFU_:b-yec0ue) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L43)):*
> ```HTML
>       var resetSpy;
> 
>       setup(function () {
> ```

Also by convention, we use the [BDD](https://mochajs.org/#bdd) (instead of TDD) interface.

---

*[test/rgb-component-input.html, line 40 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZDXgbMYcp_6SUhsPVq:-KZDXgbNt32s5EYfSGWn:b26crmy) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/rgb-component-input.html#L40)):*
> ```HTML
>     suite('basic', function () {
> 
>       test('value may be set', function () {
> ```

May set that value or it should?

---

*[test/rgb-component-validatable.html, line 12 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZkuE_vGyF7y-_6o5:-KZA3sMtVfnXyACS0RcC:buk16eb) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/rgb-component-validatable.html#L12)):*
<details><summary><i>Previously, kfox (Kelly Fox) wrote…</i></summary><blockquote>

No test required here. This is only an element used by the `rgb-component-validator` test.
</blockquote></details>

I would inline it in the related test or move it to a `fixtures` or `helpers` folder to make it obvious that it is not a test itself.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

fearphage commented:
Review status: all files reviewed at latest revision, 33 unresolved discussions.

---

*[paper-color-input.html, line 214 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7SIGj4FFugfVPywQU:-KZDYXJpSzqOUrEPQvz0:b-8neik9) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-input.html#L214)):*
<details><summary><i>Previously, kfox (Kelly Fox) wrote…</i></summary><blockquote>

I think that would add unnecessary complexity to do so. I'd have to add a new property to support that, as well as a new function to compute the value of that property, whereas the `useHexAsTitle` property is already present as part of the imported color picker behavior.

I think you can look at it like this: if we're using the hex code as the title of the color input and color picker, then the computed value of the description is just the color name. Otherwise, the description is the hex code.
</blockquote></details>

I was just saying rename the variable. The method is `_computeDescription`, but the variable is referring to the title.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

kfox commented:
Review status: all files reviewed at latest revision, 32 unresolved discussions.

---

*[paper-color-input.html, line 214 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7SIGj4FFugfVPywQU:-KZMsNgb-1krarxax_7l:b-s5hxqr) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/paper-color-input.html#L214)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

I was just saying rename the variable. The method is `_computeDescription`, but the variable is referring to the title.
</blockquote></details>

The variable name comes from `paper-color-picker-behavior`. If I change the variable name here, the definition needs to be changed there, and then the `useNameAsDescription` variable no longer "makes sense" in the `_computeTItle` function included in the behavior. If I add a `useNameAsDescription` property to this file, it would need to be a computed property that uses a function to do exactly the same thing as this function. In other words, it's just moving the "senselessness" around to a different function while adding more code. I don't see a benefit, but if you can think of another solution, let me know.

---

*[test/paper-color-circle.html, line  at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZDXt7hSrahaXteLP07:-KZMnYx3JQZOy-raTJLZ:b-patp06) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-circle.html#L)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

What's up with the empty files? Are tests missing?
</blockquote></details>

They were placeholders. There's an open issue to add these tests later.

---

*[test/paper-color-input.html, line  at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZDXx8UEhx_hw2ViTbF:-KZMnh4XmMaW-w54eiHx:b-q64pkt) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/test/paper-color-input.html#L)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

What's up with the empty files? Are tests missing?
</blockquote></details>

See above.

---

*[test/paper-color-picker-behavior.html, line 41 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9yf6Oy0UHP1Qe9zGi:-KZMnmgcWJq-YE_xnmeW:b-u00oio) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L41)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

This is the only instance that I've seen where this array is not spread across 3 lines.
</blockquote></details>

I'll change it to be consistent, and I'll move it to the `fixtures` subfolder as a separate element. I suppose I was being brief since it was a "junk" object for testing. :)

---

*[test/paper-color-picker-behavior.html, line 64 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9z-DrRuShI6Ro3v2s:-KZMnyMRXRl3VvFzy1D9:blcw4ac) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L64)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Is this necessary?
</blockquote></details>

Yes. If it is omitted, then the initial values are not set properly. The `paper-color-picker` and `paper-color-input` elements actually have listeners/observers to execute these conversions automatically. The `paper-color-picker-behavior` only provides conversion functions and common properties. I did it this way because if you share the same behavior across multiple elements and they all automatically convert values, you end up with incorrect values after those values are converted back and forth to/from RGB to hex to HSL/HSV a few times. This isn't just an issue that you can easily solve with `debounce` since each element would be firing it's own update events and forcing the other elements to convert values and fire their update events, etc. Using `debounce` just keeps that from turning into an endless loop.

---

*[test/paper-color-picker-behavior.html, line 69 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ySi_X7Ozq_tvqF06:-KZMovZNW4vXD1h-5bCq:bxixhkf) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L69)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

It would be more meaningful to me as the reader to know where this value came from (IE. do the calulation here). The same is true for all these magic values.

````
red === hex.redPart().convertedToDecimal() // pseduo code, but just to illustrate the point
```
</blockquote></details>

I'll refactor this.

---

*[test/paper-color-picker-behavior.html, line 73 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9xTZh8Lvxpqy90Zfq:-KZMp-71GfHUj2Sk_iQP:b-m6bzm0) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L73)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Where do these values come from?
</blockquote></details>

From the `tinycolor` library. Refactoring.

---

*[test/paper-color-picker-behavior.html, line 99 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9xtjNdZrfvarjnmIy:-KZMuIg3w0OL18lWkAoi:bc1kjy7) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker-behavior.html#L99)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Where's this come from? Why is it this value? This is just a random number to me. 

This assertion doesn't have any value. If I made some changes and `element.hue` was equal to `68` instead, I would have no clue why it was broken. Make these values mean something.
</blockquote></details>

These values come from the `tinycolor` library conversion functions. I've changed the comparisons to `approximately` instead of `equal`, but you're right — I should be dynamically calculating these using `tinycolor` here instead of using values that aren't meaningful.

---

*[test/paper-color-picker.html, line 30 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9sXYJcyDkQA3DAwJP:-KZMpciqvu8G2AsxMFSd:b-dpn9u8) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L30)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Add this to the `beforeEach`.
</blockquote></details>

If I do that, then the tests fail with an error: `runAfterOpen is not defined`. I'll move it inside the suite, though.

---

*[test/paper-color-picker.html, line 36 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9sRma26lrd52zMsct:-KZMqIRmfAPv_UyuvADj:b-hf550t) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L36)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

What's so slow here?
</blockquote></details>

Nothing is slow, really. I was experiencing some failures while testing where the default timeout of 10s was an unnecessary delay. I can remove this since the test shouldn't fail regularly. :)

---

*[test/paper-color-picker.html, line 43 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ8hy4_7QUNk6GWGW0Z:-KZMqt9iNPTXCtSoBjL1:b-b84xlw) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L43)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Also by convention, we use the [BDD](https://mochajs.org/#bdd) (instead of TDD) interface.
</blockquote></details>

I was following the Polymer element testing convention. (All other Polymer element tests are TDD.) We should probably talk about this.

---

*[test/rgb-component-input.html, line 40 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZDXgbMYcp_6SUhsPVq:-KZMrM_AKZFoGp5V0TYq:bqfv33p) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/rgb-component-input.html#L40)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

May set that value or it should?
</blockquote></details>

It doesn't need to be set initially; for example, the case when there is no color pre-selected and you want the user to pick a color.

---

*[test/rgb-component-validatable.html, line 12 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7ZkuE_vGyF7y-_6o5:-KZMrvHOOmW99gDd1-xN:blcchtl) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/rgb-component-validatable.html#L12)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

I would inline it in the related test or move it to a `fixtures` or `helpers` folder to make it obvious that it is not a test itself.
</blockquote></details>

Good idea. I'll move it to a `fixtures` subfolder to keep the tests themselves as small as possible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

kfox commented:
Review status: all files reviewed at latest revision, 32 unresolved discussions.

---

*[paper-color-input.html, line 95 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ7RxX1uZFqCLOhIUVx:-KZNl_PsLPLysxoA-O7q:b-5wgnb6) ([raw file](https://github.com/ustudio/paper-color-picker/blob/bee2604ccd66c289f1b17e46f92aa2d7d0d0eee3/paper-color-input.html#L95)):*
<details><summary><i>Previously, kfox (Kelly Fox) wrote…</i></summary><blockquote>

It was an overly-aggressive attempt to cleanup the code. I will add them back in.
</blockquote></details>

Actually, nevermind. I reviewed those but they were mostly incorrect/outdated and Polymer's demo/documentation helpers automatically populate the values, anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

kfox commented:
Review status: all files reviewed at latest revision, 32 unresolved discussions.

---

*[test/paper-color-picker.html, line 43 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ8hy4_7QUNk6GWGW0Z:-KZNlpOpbgNybh_UOQJ2:b-o6iu9i) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L43)):*
<details><summary><i>Previously, kfox (Kelly Fox) wrote…</i></summary><blockquote>

I was following the Polymer element testing convention. (All other Polymer element tests are TDD.) We should probably talk about this.
</blockquote></details>

After chatting with @fearphage about it, I think we'll switch to BDD-style testing since that's more likely to be supported by our dev team.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

Commit:
modified after PR feedback

Commit:
cache bower_components

fearphage commented:
Reviewed 24 of 24 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[test/hex-input.html, line 65 at r2](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZPMp31Hv2vetyAXTO8:-KZPMp31Hv2vetyAXTO9:b-nyvry1) ([raw file](https://github.com/ustudio/paper-color-picker/blob/e6dff0e2f280960ac987154616a3304bbcb9d273/test/hex-input.html#L65)):*
> ```HTML
>         var element = fixture('basic');
>         element.value = 'foobar';
>         expect(element.value).to.equal(element.inputElement.value);
> ```

MINOR: Does `element.inputElement.value` equal the string `foobar` like the input? If so, I would use that in the assertion instead.

---

*[test/hex-input.html, line 115 at r2](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZPNNCplIZvLMiKunSn:-KZPNNCplIZvLMiKunSo:b7i46le) ([raw file](https://github.com/ustudio/paper-color-picker/blob/e6dff0e2f280960ac987154616a3304bbcb9d273/test/hex-input.html#L115)):*
> ```HTML
>       });
> 
>       it('should change the invalid attribute on value changes', function() {
> ```

MINOR: `should update` reads a little better.

---

*[test/paper-color-picker-behavior.html, line 47 at r2](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZPOEXEkeq3tAM3uCA_:-KZPOEXF5MS0-3CvhAnl:b-6bsupb) ([raw file](https://github.com/ustudio/paper-color-picker/blob/e6dff0e2f280960ac987154616a3304bbcb9d273/test/paper-color-picker-behavior.html#L47)):*
> ```HTML
>     describe('basic functionality', function () {
> 
>       function colorTest (element, color) {
> ```

Thanks. This is so much better. :star2: 

---

*[test/paper-color-picker.html, line 30 at r1](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZ9sXYJcyDkQA3DAwJP:-KZPOycHPuZJ2hzi2RUN:b-sgxc65) ([raw file](https://github.com/ustudio/paper-color-picker/blob/19ae9eac7741d95fc212d9b609281f654e00a13c/test/paper-color-picker.html#L30)):*
<details><summary><i>Previously, kfox (Kelly Fox) wrote…</i></summary><blockquote>

If I do that, then the tests fail with an error: `runAfterOpen is not defined`. I'll move it inside the suite, though.
</blockquote></details>

I was saying take the 2 lines in `runAfterOpen` and put them in the `beforeEach`. Then the `callback` would be `function() { done(); }`. This is not blocking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

tjensen commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-:-KZRmlhYbGq5zRr5Ly8m:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tjensen commented:
Reviewed 1 of 24 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

kfox commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/hex-input.html, line 65 at r2](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZPMp31Hv2vetyAXTO8:-KZRsZuMEiotkgHM9-O0:b-xhd7h6) ([raw file](https://github.com/ustudio/paper-color-picker/blob/e6dff0e2f280960ac987154616a3304bbcb9d273/test/hex-input.html#L65)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

MINOR: Does `element.inputElement.value` equal the string `foobar` like the input? If so, I would use that in the assertion instead.
</blockquote></details>

Yep, I'll change that.

---

*[test/hex-input.html, line 115 at r2](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZPNNCplIZvLMiKunSn:-KZRsdzVqOOA2pAiOG9r:b-pbdqz3) ([raw file](https://github.com/ustudio/paper-color-picker/blob/e6dff0e2f280960ac987154616a3304bbcb9d273/test/hex-input.html#L115)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

MINOR: `should update` reads a little better.
</blockquote></details>

Agreed.

---

*[test/paper-color-picker-behavior.html, line 47 at r2](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1#-KZPOEXEkeq3tAM3uCA_:-KZRsgAIxuATRgBGlDDZ:b-8fx2u7) ([raw file](https://github.com/ustudio/paper-color-picker/blob/e6dff0e2f280960ac987154616a3304bbcb9d273/test/paper-color-picker-behavior.html#L47)):*
<details><summary><i>Previously, fearphage (Phred) wrote…</i></summary><blockquote>

Thanks. This is so much better. :star2: 
</blockquote></details>

:+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->

Commit:
modified after PR feedback

fearphage commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ustudio/paper-color-picker/1)*
<!-- Sent from Reviewable.io -->
